### ISSUE TITLE
Applying transforms to animate in new IGListSectionControllers

### ISSUE THREAD

**Issue by xiaosongshu** [Instagram/IGListKit#708]
**[Background]**
Was looking at @rnystrom comment in #629 where he mentioned cell transforms could be applied in IGListDisplayDelegate's `willDisplay` method. I'm also trying to apply cell transforms, but my use case is slightly different, as I have two arrays, one being messages already displayed to the user, and one for new messages that haven't been seen yet. For each new message that hasn't been seen yet, I want to animate each one in with a scale transform, one after the other with a slight delay between each. Eventually the messages will be removed from the new message array, and will each be appended to the existing messages array.

I got this working previously by using a pretty complicated array of "pendingMessages", where one message would be animated at a time via `performBatchUpdates`, and in the completion block, I would call an `animate` method on the cell to perform the transforms, then recursively call `performBatchUpdates` to animate in the next "pendingMessage." This gave me plenty of headaches, which lead to me discovering IGListKit :)

```swift
// animatePendingMessages() {
// ...
//    self.collectionView.performBatchUpdates({ () -> Void in
// ...
        }, completion: { finish in
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                if finish {
                    self.collectionView.scrollToItem(at: IndexPath(row: self.returnDisplayedMessages().count - 1, section: 0), at: .top, animated: true)
                    if let cell = (self.collectionView.cellForItem(at: IndexPath(row: self.returnDisplayedMessages().count - 1, section: 0)) as? MessageCollectionViewCell) {
                        cell.animate { // perform the transforms
                            animateNext() // animate the next message in "pendingMessages" array
                        }
                    } else {
                        animateNext()
                    }
                }
            }
```

You can see what the animate method inside the MessageCollectionViewCell looks like here:

```swift
func animate(completion: (() -> Void)?) {
        topView.isHidden = true
        let resizeAnimation = CABasicAnimation(keyPath: "transform.scale")
        resizeAnimation.fromValue = 0.0
        resizeAnimation.toValue = 1.0
        resizeAnimation.duration = 0.4
        resizeAnimation.beginTime = CACurrentMediaTime()
        resizeAnimation.fillMode = kCAFillModeBackwards
        messageLabel.layer.add(resizeAnimation, forKey: "scale")
        containerView.layer.add(resizeAnimation, forKey: "scale")
        coverView.layer.add(resizeAnimation, forKey: "scale")
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.4) {
            completion?()
        }
    }
```

The problem I'm having is adding back this `cell.animate` method in IGListDisplayDelegate methods. Adding it to `willDisplay` animates every cell (IGListSection) every time you scroll which I don't want (since I only want to animate in the new messages once), whereas calling it in `didEndDisplaying` never actually shows the animations.

```swift
    func listAdapter(_ listAdapter: IGListAdapter, didEndDisplaying sectionController: IGListSectionController, cell: UICollectionViewCell, at index: Int) {
        if let cell = cell as? MessageCollectionViewCell {
            cell.animate(completion: nil)
        }
    }
```

Also, not sure if relevant, but currently I actually don't see any animations at all when I call ```self.adapter.performUpdates(animated: true)```. The cells just appear instantly. Wasn't sure if maybe I was missing something or if the animations should in fact come out of the box

**[Question]**
Do you have any idea how I can go about reusing my `cell.animate` method call, to animate in only the new messages, one after each other? So effectively you can imagine the collectionView first scrolls to the bottom, then one new message animates in after each other (think similar to the Qu@rtz news app). I would have imagined I could have just used `didEndDisplaying` method, but didn't have much luck with that unfortunately and adding to `willDisplay` unfortunately fires for each cell every single time as you scroll which I don't want.

So in short it's:
1. Where should I be applying transforms to SectionControllers
2. Where should I be applying delays, so that cells animate one after each other (rather than all appearing at once)
`DispatchQueue.main.asyncAfter(deadline: .now() + 0.3)`

**Comment by rnystrom** [comment]
Ok I think I understand your setup:

- You call `performUpdates(...)` with a subset of all messages
- You then append a new message, call `performUpdates(...)` again, then animate in the cell
- In the completion block of the animation, you start over again
- Rinse & repeat until all cells have animated in

Is that right?

If that's the case, then you should be able to do something similar using the completion block in `performUpdates(...)`:

```swift
func animatePendingMessages() {
  adapter.performUpdates(animated: true) {
    let lastObjectIndex = adapter.objects.count - 1
    let path = IndexPath(item: 0, section: lastObjectIndex)
    if let cell = self.collectionView.cellForItem(at: path) as? MessageCollectionViewCell {
        cell.animate {
            animateNext()
        }
    } else {
        animateNext()
    }
  }
}
```

> Where should I be applying transforms to SectionControllers

This is a unique type of animation, your design here might be the best. There's probably a clean way to abstract this into its own object as well.

> Where should I be applying delays, so that cells animate one after each other (rather than all appearing at once) `DispatchQueue.main.asyncAfter(deadline: .now() + 0.3)`

Again, probably best all bundled into one function/object that handles all of this for you.

**Comment by xiaosongshu** [comment]
Hi @rnystrom, thanks for the suggestion and apologies for the delayed response! Checked back repeatedly first few days but didn't see a reply, and later was looking out for a github email notification to only realize my notifications aren't working anymore.

Gave your suggestion a shot but unfortunately am still struggling with this. After sending a new message, the collection view seems to update instantly (old messages removed instantaneously, and then the new messages all appear at once instantaneously as well, however just the last message has the animation). My guess is the animations for the earlier inserted messages are firing, but then the subsequent calls to performUpdate are canceling the previous animations? Is there a way to delay the insertions _and_ deletions of sections until previous sections finish animating? If not, I'm guessing I might have to rebuild the `pendingMessages` array again and remove pendingMessages one by one, each time calling the next `performUpdates` only after the previous `performUpdates` finishes. Let me know if you think there might be an easier way as the `pendingMessages` array is quite complex given it stores both messages to delete and messages to insert (and lead to too many index out of bound exceptions which was why I was hoping IGListKit could come to the rescue).

What you outlined was exactly how the old setup worked. To clarify how I have the new setup with IGListKit, I've created a simplified example below:

1. Method called `start` fires, which calls `removeLastFewMessages ` followed by `insertMultipleMessages`
1. Method `insertMultipleMessages` calls `insertLast` multiple times, with the new message + some autogenerated messages
2. Each `insertLast` calls `adapter.performUpdates`
3. Expected: each new message should animate in (simple scale transform / "grow" effect), one after each other
4. What's happening: all deleted messages disappear instantaneously & new messages appear simultaneously with only the last message having the transform animation


```
    func start() {
        self.removeLastFewMessages()    
        self.insertMultipleMessages()
    }
```


```
    func insertMultipleMessages() {
        self.insertLast(with message: message_new)      
        self.insertLast(with message: message_generated_1)      
        self.insertLast(with message: message_generated_2)      
    }
```

```
    func insertLast(with message: Message) {
        
        self.adapter.performUpdates(animated: true) { (finish) in
            if finish {
                
                let lastObjectIndex = self.adapter.objects().count - 1
                let path = IndexPath(item: 0, section: lastObjectIndex)
                if let cell = self.collectionView.cellForItem(at: path) as? MessageCollectionView {
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                        cell.animate(completion: {
                            self.collectionView.scrollToBottom()
                        })
                    }
            }
        }
    }
```

**Comment by rnystrom** [comment]
Gonna close this out after reviewing. Old issue without a lot of activity, and I'm still not sure where exactly the problem lies. Feel free to comment if you have it working or want to keep this rolling.
